flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -75,10 +137,281 @@ public DefaultScheduler(
 			slotRequestTimeout,
 			shuffleMaster,
 			partitionTracker);
+
+		this.restartBackoffTimeStrategy = restartBackoffTimeStrategy;
+		this.slotRequestTimeout = slotRequestTimeout;
+		this.slotProvider = slotProvider;
+		this.delayExecutor = delayExecutor;
+		this.userCodeLoader = userCodeLoader;
+		this.schedulingStrategyFactory = checkNotNull(schedulingStrategyFactory);
+		this.failoverStrategyFactory = checkNotNull(failoverStrategyFactory);
+		this.executionVertexOperations = checkNotNull(executionVertexOperations);
+		this.executionVertexVersioner = executionVertexVersioner;
+		this.conditionalFutureHandlerFactory = new ConditionalFutureHandlerFactory(executionVertexVersioner);
+	}
+
+	// ------------------------------------------------------------------------
+	// SchedulerNG
+	// ------------------------------------------------------------------------
+
+	@Override
+	public void startSchedulingInternal() {
+		initializeScheduling();
+		schedulingStrategy.startScheduling();
+	}
+
+	private void initializeScheduling() {
+		executionFailureHandler = new ExecutionFailureHandler(failoverStrategyFactory.create(getFailoverTopology()), restartBackoffTimeStrategy);
+		schedulingStrategy = schedulingStrategyFactory.createInstance(this, getSchedulingTopology(), getJobGraph());
+		executionSlotAllocator = new DefaultExecutionSlotAllocator(slotProvider, getInputsLocationsRetriever(), slotRequestTimeout);
+		setTaskFailureListener(new UpdateTaskExecutionStateInDefaultSchedulerListener(this, getJobGraph().getJobID()));
+		prepareExecutionGraphForScheduling();
+	}
+
+	@Override
+	public boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
+		final Optional<ExecutionVertexID> executionVertexIdOptional = getExecutionVertexId(taskExecutionState.getID());
+		if (executionVertexIdOptional.isPresent()) {
+			final ExecutionVertexID executionVertexId = executionVertexIdOptional.get();
+			updateState(taskExecutionState);
+			schedulingStrategy.onExecutionStateChange(executionVertexId, taskExecutionState.getExecutionState());
+			maybeHandleTaskFailure(taskExecutionState, executionVertexId);
```

Comment:
Can it happen that we transition due to `schedulingStrategy.onExecutionStateChange` into `ExecutionState.FAILED` which is not handled by the `onExecutionStateChange` call?

Reply:
I don't see it but maybe I am wrong.

Reply:
Failures are not handled in `LazyFromSourcesSchedulingStrategy` or `EagerSchedulingStrategy`. The scheduler is responsible for handling task failures.

Reply:
@zhuzhurk These are valid points.

If I understand you correctly, you are proposing to wrap the body of `allocateSlotsAndDeploy()` in

```
getMainThreadExecutor().execute(() -> {
...
}
```

Is that right?

Reply:
Exactly.

Reply:
@zhuzhurk

I gave this some thought. 

> when we invoke maybeHandleTaskFailure right after invoking schedulingStrategy.onExecutionStateChange, the task state may even have changed in the call stack chain so that we are doing failover handling in an unexpected state

If we added `getMainThreadExecutor().execute()` to `allocateSlotsAndDeploy()`, I think it wouldn't solve this issue. We still might re-deploy tasks that were already re-deployed. 

I think there is a deeper underlying problem to why these lines look confusing:
```
schedulingStrategy.onExecutionStateChange(executionVertexId, taskExecutionState.getExecutionState());
maybeHandleTaskFailure(taskExecutionState, executionVertexId);
```

The `SchedulingStrategy` gets informed about task failures but is not supposed to re-deploy or reason about failed tasks because the failure handling is done centrally (e.g. inside `flip1.FailoverStrategy`). 

At the moment, I wouldn't change anything.

Reply:
I think it's fine to not change it at the moment, given that 
1. At the moment, the schedulingStrategy.onExecutionStateChange() is only notified for FAILED state internally and RUNNING/FINISHED/CANCELED/FAILED state from TM side 
2. RUNNING/FAILED/CANCELED/FAILED state do not tigger scheduling in the eager/lazy scheduling strategies
3. FINISHED state is currently used for lazy scheduling but it may be changed as discussed [in this comment](https://github.com/apache/flink/pull/9663#discussion_r326540913) 

We can reconsider this when later we are to redefine the scheduling strategy interfaces or refine the invocations as discussed in [the other comment](https://github.com/apache/flink/pull/9663/files#r326897974).